source/bringauto/modules/ModuleManagerLibraryHandlerAsync.cpp

```diff
@@ -0,0 +1,211 @@
+#include <bringauto/modules/ModuleManagerLibraryHandlerAsync.hpp>
+#include <bringauto/settings/Constants.hpp>
+
+#include <fleet_protocol/common_headers/general_error_codes.h>
+
+#include <csignal>
+
+
+
+namespace bringauto::modules {
+
+async_function_execution::AsyncFunctionExecutor aeronClient {
+	async_function_execution::Config {
+		.isProducer = true,
+		.defaultTimeout = std::chrono::seconds(1)
+	},
+	async_function_execution::FunctionList {
+		getModuleNumberAsync,
+		isDeviceTypeSupportedAsync,
+		sendStatusConditionAsync,
+		generateCommandAsync,
+		aggregateStatusAsync,
+		aggregateErrorAsync,
+		generateFirstCommandAsync,
+		statusDataValidAsync,
+		commandDataValidAsync
+	}
+};
+
+ModuleManagerLibraryHandlerAsync::ModuleManagerLibraryHandlerAsync(const std::string &moduleBinaryPath) :
+		moduleBinaryPath_ { moduleBinaryPath } {
+	aeronClient.connect();
+	deallocate_ = [this](struct buffer *buffer) {
+		this->deallocate(buffer);
+	};
+}
+
+ModuleManagerLibraryHandlerAsync::~ModuleManagerLibraryHandlerAsync() {
+	if (moduleBinaryProcess_.valid()) {
+		::kill(moduleBinaryProcess_.id(), SIGTERM);
+		moduleBinaryProcess_.wait();
+	}
+}
+
+void ModuleManagerLibraryHandlerAsync::loadLibrary(const std::filesystem::path &path) {
+	moduleBinaryProcess_ = boost::process::child { moduleBinaryPath_, "-m", path.string() };
+	if (!moduleBinaryProcess_.valid()) {
+		throw std::runtime_error { "Failed to start module binary " + moduleBinaryPath_ };
+	}
+	std::this_thread::sleep_for(std::chrono::seconds(1)); // TODO Not sure how much time is needed.
+}
+
+int ModuleManagerLibraryHandlerAsync::getModuleNumber() const {
+	return aeronClient.callFunc(getModuleNumberAsync).value();
+}
+
+int ModuleManagerLibraryHandlerAsync::isDeviceTypeSupported(unsigned int device_type) {
+	std::lock_guard lock { tmpMutex_ };
+	return aeronClient.callFunc(isDeviceTypeSupportedAsync, device_type).value();
+}
+
+int ModuleManagerLibraryHandlerAsync::sendStatusCondition(const Buffer &current_status,
+														  const Buffer &new_status,
+													 	  unsigned int device_type) const {
+	ConvertibleBuffer current_status_raw_buffer;
+	ConvertibleBuffer new_status_raw_buffer;
+
+	if (current_status.isAllocated()) {
+		current_status_raw_buffer = current_status.getStructBuffer();
+	}
+	if (new_status.isAllocated()) {
+		new_status_raw_buffer = new_status.getStructBuffer();
+	}
+
+	return aeronClient.callFunc(sendStatusConditionAsync, current_status_raw_buffer, new_status_raw_buffer, device_type).value();
+}
+
+int ModuleManagerLibraryHandlerAsync::generateCommand(Buffer &generated_command,
+													  const Buffer &new_status,
+													  const Buffer &current_status,
+													  const Buffer &current_command, unsigned int device_type) {
+	ConvertibleBuffer new_status_raw_buffer;
```

Comment:
Are we sure we do not need a mutex for functions which can be called in parallel? For example, if multiple devices are connected, etc.?